### PR TITLE
Add a default serializer

### LINKED ISSUE #191: Expose default serializer
**Is your feature request related to a problem? Please describe.**
Cosmos REST contracts are JSON.net annotated. For customers using streaming API a serialization implementation which can work with them is required.

**Describe the solution you'd like**
CosmosJsonSerializer.Default

**Describe alternatives you've considered**
Custom JSON.net code but it breaks once moved out of JSON.net FX.

### PR DESCRIPTION
**Pull request by j82w**
# Pull Request Template

## Description
Adding a default serializer to CosmosClient. This will allow users to deserialize meta data stream responses like CosmosDatabaseSettings.

- The default serializer will always be used for meta data types like Database and Container operations. This is because they are system owned types with specific Newtonsoft tags to be serialized correctly.
- All item operations including query will use the user specified version when it is available. If no user version is set it will use the default implementation.
- It is a instance variable on CosmosClient to allow unit testing. 

## Type of change

Please delete options that are not relevant.
- [x] New feature (non-breaking change which adds functionality)

## Closing issues

closes #191

### CODE REVIEW

File: Microsoft.Azure.Cosmos/src/CosmosClient.cs [hunk 1]
```diff
@@ -205,14 +205,21 @@ internal CosmosClient(CosmosClientConfiguration cosmosClientConfiguration)
         /// </summary>
         public virtual CosmosClientConfiguration Configuration { get; private set; }
 
+        /// <summary>
+        /// The default Cosmos JSON serializer that is used for meta data types like <see cref="CosmosContainerSettings"/>
+        /// </summary>
+        public virtual CosmosJsonSerializer DefaultJsonSerializer { get; private set; }
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/CosmosClient.cs]: Can we please keep this internal until an explicit user ask?
All client configuration is modeled as options, may be that's the right place for this.

Comment (j82w) [reply Microsoft.Azure.Cosmos/src/CosmosClient.cs]: Fixed in latest

File: Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs [hunk 2]
```diff
@@ -238,15 +239,29 @@ internal set
...
+            internal set => this.userJsonSerializer = value ?? throw new NullReferenceException(nameof(this.UserJsonSerializer));
         }
 
+        /// <summary>
+        /// A JSON serializer used by the CosmosClient to serialize or de-serialize cosmos request/responses.
+        /// The default serializer is always used for all system owned types like CosmosDatabaseSettings.
+        /// The default serializer is used for user types if no UserJsonSerializer is specified
+        /// </summary>
+        [JsonConverter(typeof(ClientConfigurationJsonConverter))]
+        internal virtual CosmosJsonSerializer DefaultJsonSerializer => CosmosClientOptions.defaultJsonSerializer;
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs]: SettingsJsonSerializer.

File: Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs [hunk 3]
```diff
@@ -238,15 +239,29 @@ internal set
 
         /// <summary>
         /// A JSON serializer used by the CosmosClient to serialize or de-serialize cosmos request/responses.
-        /// If no custom JSON converter was set it uses the default <see cref="CosmosDefaultJsonSerializer"/>
+        /// If no custom JSON converter was set it uses the default <see cref="CosmosJsonSerializerCore"/>
         /// </summary>
         [JsonConverter(typeof(ClientConfigurationJsonConverter))]
-        public virtual CosmosJsonSerializer CosmosJsonSerializer
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs]: CosmosJsonSerializer with older logic.

File: Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs (hunk 2 above)
Comment (j82w) [reply Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs]: As discussed used SettingsSerializer

File: Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs (hunk 3 above)
Comment (j82w) [reply Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs]: It currently returns it in the wrapper which is internal so users can't do anything with it. By having the separate method users can get access to the original object they set, and makes it useful for different user scenarios.

**Pull request merged**